tests/common/reboot.py

```diff
@@ -0,0 +1,99 @@
+import time
+import logging
+from multiprocessing.pool import ThreadPool, TimeoutError
+from ansible_host import AnsibleModuleException
+
+logger = logging.getLogger(__name__)
+
+# SSH defines
+SONIC_SSH_PORT  = 22
+SONIC_SSH_REGEX = 'OpenSSH_[\\w\\.]+ Debian'
+
+# map reboot type -> reboot command
+reboot_commands =\
+{
+    'cold': 'reboot',
+    'fast': 'fast-reboot',
+    'warm': 'warm-reboot',
+}
+
+
+def reboot(duthost, localhost, reboot_type='cold', delay=10, timeout=180, wait=120):
+    """
+    reboots DUT
+    :param duthost: DUT host object
+    :param localhost:  local host object
+    :param reboot_type: reboot type (cold, fast, warm)
+    :param delay: delay between ssh availability checks
+    :param timeout: timeout for waiting ssh port state change
+    :param wait: time to wait for DUT to initialize
+    :return:
+    """
+
+    # pool for executing tasks asynchronously
+    pool = ThreadPool()
+
+    try:
+        reboot_command = reboot_commands[reboot_type]
+    except KeyError:
+        raise ValueError('invalid reboot type: "{}"'.format(reboot_type))
+
+    def execute_reboot():
+        logger.info('rebooting with command "{}"'.format(reboot_command))
+        return duthost.command(reboot_command)
+
+    reboot_res = pool.apply_async(execute_reboot)
+
+    logger.info('waiting for ssh to drop')
+    res = localhost.wait_for(host=duthost.hostname,
+                             port=SONIC_SSH_PORT,
+                             state='absent',
+                             search_regex=SONIC_SSH_REGEX,
+                             delay=delay,
+                             timeout=timeout)
+
+    if 'failed' in res['localhost']:
+        if reboot_res.ready():
+            logger.error('reboot result: {}'.format(reboot_res.get()))
+        raise Exception('DUT did not shutdown')
```

Comment:
In recent testing, I found that on some testbed, rebooting happens fast. The wait_for module may not be able to detect the switch down event and timeout eventually. But indeed the switch rebooting was successful. I think it's better to ignore the failure of the wait_for down check and use uptime to verify that the switch was indeeded rebooted. For example:
1. Get date and time on DUT before the reboot command is issued.
2. After switch is up again (wait_for up succeeded), run "uptime -s" to get a new datetime. Assert that this new datetime is at least for example 10 seconds later than the previously recorded datetime.

Reply:
It's surprising to me that some SONiC system can reboot so fast.
Note, we execute reboot command in parallel to listening to SSH port. Assuming they start almost at the same time the reboot script itself has some serivces to stop, executes 'sleep 3', etc.
So reboot command definitely takes time (>3sec) by that time I think it is safe to assume wait_for routine has started. Even if reboot command execution time -> 0, the time it takes from grub to kernel to ssh service start is not 0,

Reply:
Theoretically it is correct. But I suspect there is other issue caused the host executing the wait_for ansible module not able to detect the SSH down status in time. Could you try the code on our testbed "mts-sonic-dut"?